src/routers/alpha-router/alpha-router.ts

```diff
@@ -1109,23 +1111,38 @@ export class AlphaRouter
       if (!this.simulator) {
         throw new Error('Simulator not initialized!');
       }
-      const beforeSimulate = Date.now();
-      const swapRouteWithSimulation = await this.simulator.simulateTransaction(
-        swapConfig.simulate.fromAddress,
-        swapRoute,
-        this.l2GasDataProvider
-          ? await this.l2GasDataProvider!.getGasData()
-          : undefined
-      );
-      metric.putMetric(
-        'SimulateTransaction',
-        Date.now() - beforeSimulate,
-        MetricLoggerUnit.Milliseconds
-      );
-      return swapRouteWithSimulation;
+      const fromAddress = swapConfig.simulate.fromAddress;
+      if (
+        await this.userHasSufficientBalance(
+          fromAddress,
+          tradeType,
+          amount,
+          // Quote will be in WETH even if quoteCurrency is ETH
+          // So we init a new CurrencyAmount object here
+          CurrencyAmount.fromRawAmount(quoteCurrency, quote.quotient.toString())
+        )
+      ) {
```

Comment:
Why not make this part of the simulator class itself? As written we make an assumption that no simulator can ever succeed unless the from address has balance, but I may implement an `ISimulator` which can actually handle this case (maybe i transfer tokens to the from address on a fork, or something like that)

Reply:
moved it to the simulator. Let me know if i should make it a static method, i was not sure.

Reply:
Also, I changed ISimulator to an abstract class to handle this